In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

import sklearn
print(f"Using Scikit-Learn version: {sklearn.__version__}")

Using Scikit-Learn version: 1.5.1


## 0. An end-to-end Scikit-Learn workflow

Before we get in-depth, let's quickly check out what an end-to-end Scikit-Learn workflow might look like.

Once we've seen an end-to-end workflow, we'll dive into each step a little deeper.

Specifically, we'll get hands-on with the following steps:

   1. Getting data ready (split into features and labels, prepare train and test steps)
   2. Choosing a model for our problem
   3. Fit the model to the data and use it to make a prediction
   4. Evaluate the model
   5. Experiment to improve
   6. Save a model for someone else to use


## Random Forest Classifier Workflow for Classifying Heart Disease

### 1. Get the data ready

As an example dataset, we'll import heart-disease.csv.

This file contains anonymised patient medical records and whether or not they have heart disease or not (this is a classification problem since we're trying to predict whether something is one thing or another).

In [2]:
import pandas as pd
heart_disease = pd.read_csv("../data/heart-disease.csv")
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Here, each row is a different patient and all columns except `target` are different patient characteristics.
The `target` column indicates whether the patient has heart disease (target=1) or not (target=0), this is our "label" columnm, the variable we're going to try and predict.

The rest of the `columns` (often called `features`) are what we'll be using to predict the target value.

In [3]:
# Create X (all the feature columns)
X = heart_disease.drop("target", axis=1)

# Create y (the target column)
y = heart_disease["target"]

# Check the head of the features DataFrame
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [4]:
#check the head and value counts of the labels
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [5]:
y.value_counts()

target
1    165
0    138
Name: count, dtype: int64

In [6]:
y

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64



One of the most important practices in machine learning is to split datasets into `training and test sets.`

As in, a model will train on the training set to learn patterns and then those patterns can be evaluated on the test set.

Crucially, a model should never see testing data during training.

This is equivalent to a student studying course materials during the semester (training set) and then testing their abilities on the following exam (testing set).

Scikit-learn provides the `sklearn.model_selection.train_test_split` method to split datasets in training and test sets.

In [7]:
#split the data into training and test sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((227, 13), (76, 13), (227,), (76,))


## 2. Choose the model and hyperparameters

Choosing a model often depends on the type of problem you're working on.

For example, there are different models that Scikit-Learn recommends whether you're working on a classification or regression problem.

You can see a map breaking down the different kinds of `model` options and recommendations in the Scikit-Learn documentation.

Scikit-Learn refers to models as "estimators", however, they are often also referred to as model or `clf` (short for classifier).

A model's hyperparameters are settings you can change to adjust it for your problem, much like knobs on an oven you can tune to cook your favourite dish.


In [8]:
# since we're working on classification problem, we'll start with a RandomForestClassification
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier()

we can see the current hyperparameter of a model with the `get_params()` method

In [9]:
# View the current hyperparameters
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}


### 3. Fit the model to the data and use it to make a prediction

Fitting a model a dataset involves passing it the data and asking it to figure out the patterns.

If there are labels (supervised learning), the model tries to work out the relationship between the data and the labels.

If there are no labels (unsupervised learning), the model tries to find patterns and group similar samples together.

Most Scikit-Learn models have the fit(X, y) method built-in, where the X parameter is the features and the y parameter is the labels.

In our case, we start by fitting a model on the training split (X_train, y_train).


In [10]:
clf.fit(X=X_train, y=y_train)

RandomForestClassifier()

### Use the model to make a prediction

The whole point of training a machine learning model is to use it to make some kind of prediction in the future.

Once your model instance is trained, you can use the predict() method to predict a target value given a set of features.

In other words, use the model, along with some new, unseen and unlabelled data to predict the label.

In [12]:
# This doesn't work... incorrect shapes
array=np.array([0,2,3,4])
array.shape
y_label = clf.predict(np.array([0,2,3,4]))

/home/caro/Desktop/simple_project/env/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


ValueError: Expected 2D array, got 1D array instead:
array=[0. 2. 3. 4.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

We get a ValueError (mismatched shapes):

ValueError: Expected 2D array, got 1D array instead:
array=[0. 2. 3. 4.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

This happens because we're trying to make predictions on data that is in a different format to the data our model was trained on.

Since our model was trained on data from X_train, predictions should be made on data in the same format and shape as X_train.

Our goal in many machine learning problems is to use patterns learned from the training data to make predictions on the test data (or future unseen data).

In [13]:
# In order to predict a label, data has to be in the same shape as X_train
X_test.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
117   56    1   3       120   193    0        0      162      0      1.9   
290   61    1   0       148   203    0        1      161      0      0.0   
52    62    1   2       130   231    0        1      146      0      1.8   
155   58    0   0       130   197    0        1      131      0      0.6   
151   71    0   0       112   149    0        1      125      0      1.6   

     slope  ca  thal  
117      1   0     3  
290      2   1     3  
52       1   3     3  
155      1   0     2  
151      1   0     2

In [14]:
# use the model to make a prediction on the test data (further evaluation)
y_preds = clf.predict(X=X_test)


### 4. Evaluate the model

Now we've made some predictions, we can start to use some more Scikit-Learn methods to figure out how good our model is.

Each model or estimator has a built-in `score()` method.

This method compares how well the model was able to learn the patterns between the features and labels.

The score() method for each model uses a standard evaluation metric to measure your model's results.

In the case of a classifier (our model), one of the most common evaluation metrics is accuracy (the fraction of correct predictions out of total predictions).

Let's check out our model's accuracy on the training set.


In [15]:
# Evaluate the model on the training set
train_acc = clf.score(X=X_train, y= y_train)
print(f"The model's accuracy on the training dataset is: {train_acc*100}%")

The model's accuracy on the training dataset is: 100.0%




The model's accuracy on the training dataset is: 100.0%

Woah! Looks like our model does pretty well on the training datset.

This is because it has a chance to see both data and labels.

How about the test dataset?


In [16]:
# Evaluate the model on the test set
test_acc = clf.score(X=X_test, y= y_test)
print(f"The model's accuracy on the testing dataset is: {test_acc*100:.2f}%")

The model's accuracy on the testing dataset is: 82.89%




The model's accuracy on the testing dataset is: 88.16%

Hmm, looks like our model's accuracy is a bit less on the test dataset than the training dataset.

This is quite often the case, because remember, a model has never seen the testing examples before.

There are also a number of other evaluation methods we can use for our classification models.

All of the following classification metrics come from the sklearn.metrics module:

`classification_report(y_true, y_true)` - Builds a text report showing various classification metrics such as precision, recall and F1-score.
`confusion_matrix(y_true, y_pred)` - Create a confusion matrix to compare predictions to truth labels.
`accuracy_score(y_true, y_pred)` - Find the accuracy score (the default metric) for a classifier.

All metrics have the following in common: they compare a model's predictions (y_pred) to truth labels (y_true).


In [17]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

#create a classification report
print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.83      0.75      0.79        32
           1       0.83      0.89      0.86        44

    accuracy                           0.83        76
   macro avg       0.83      0.82      0.82        76
weighted avg       0.83      0.83      0.83        76



In [18]:
# Create a confusion matrix
conf_mat = confusion_matrix(y_test, y_preds)
conf_mat

array([[24,  8],
       [ 5, 39]])

In [19]:
# Compute the accuracy score (same as the score() method for classifiers) 
accuracy_score(y_test, y_preds)

0.8289473684210527

## 5. Experiment to improve

The first model you build is often referred to as a baseline (a baseline is often even simpler than the model we've used, a baseline could be "let's just by default predict the most common value and then try to improve").

Once you've got a baseline model, like we have here, it's important to remember, this is often not the final model you'll use.

The next step in the workflow is to try and improve upon your baseline model.

How?

With one of the most important mottos in machine learning...

Experiment, experiment, experiment!

Experiments can come in many different forms.

But let's break it into two.

    1. From a model perspective.
    2. From a data perspective.

From a model perspective may involve things such as using a more complex model or tuning your models hyperparameters.

From a data perspective may involve collecting more data or better quality data so your existing model has more of a chance to learn the patterns within.

If you're already working on an existing dataset, it's often easier try a series of model perspective experiments first and then turn to data perspective experiments if you aren't getting the results you're looking for.

One thing you should be aware of is if you're tuning a models hyperparameters in a series of experiments, your reuslts should always be cross-validated (we'll see this later on!).

Cross-validation is a way of making sure the results you're getting are consistent across your training and test datasets (because it uses multiple versions of training and test sets) rather than just luck because of the order the original training and test sets were created.

   * Try different hyperparameters.
   * All different parameters should be cross-validated.
       * Note: Beware of cross-validation for time series problems (as for time series, you don't want to mix samples from the future with samples from the past).

Different models you use will have different hyperparameters you can tune.

For the case of our model, the `RandomForestClassifier()`, we'll start trying different values for `n_estimators` (a measure for the number of trees in the random forest).

By default, n_estimators=100, so how about we try values from 100 to 200 and see what happens (generally more is bet

ter)?


In [20]:
# Try different numbers of estimators (trees)... (no cross-validation)
np.random.seed(42)
for i in range(100, 200 , 10):
    print(f'trying model with {i} estimators ... ')
    model = RandomForestClassifier(n_estimators=i).fit(X_train, y_train)
    print(f"Model accuracy on test set : {model.score(X_test, y_test) * 100:.2f}%")
    print("")

trying model with 100 estimators ... 
Model accuracy on test set : 86.84%

trying model with 110 estimators ... 
Model accuracy on test set : 86.84%

trying model with 120 estimators ... 
Model accuracy on test set : 86.84%

trying model with 130 estimators ... 
Model accuracy on test set : 85.53%

trying model with 140 estimators ... 
Model accuracy on test set : 85.53%

trying model with 150 estimators ... 
Model accuracy on test set : 86.84%

trying model with 160 estimators ... 
Model accuracy on test set : 86.84%

trying model with 170 estimators ... 
Model accuracy on test set : 85.53%

trying model with 180 estimators ... 
Model accuracy on test set : 84.21%

trying model with 190 estimators ... 
Model accuracy on test set : 86.84%





The metrics above were measured on a single train and test split.

Let's use `sklearn.model_selection.cross_val_score` to measure the results across 5 different train and test sets.

We can achieve this by setting `cross_val_score(X, y, cv=5).`

Where X is the full feature set and y is the full label set and cv is the number of train and test splits cross_val_score will automatically create from the data (in our case, 5 different splits, this is known as 5-fold cross-validation).


In [21]:
from sklearn.model_selection import cross_val_score

# with cross-validation
np.random.seed(42)
for i in range(100, 200, 10):
    print(f"trying model with {i} estimators...")
    model = RandomForestClassifier(n_estimators=i).fit(X_train, y_train)

    # Measure the model score on a single train/test split
    model_score = model.score(X_test, y_test)
    print(f"Model accuracy on single test set split : {model_score * 100:.2f}%")

    # Measure the mean cross-validation score across 5 different train and test splits
    cross_val_mean = np.mean(cross_val_score(model, X, y, cv=5))
    print(f"5-fold cross-validation score: {cross_val_mean * 100:.2f}%")

    print("")

trying model with 100 estimators...
Model accuracy on single test set split : 86.84%
5-fold cross-validation score: 82.15%

trying model with 110 estimators...
Model accuracy on single test set split : 85.53%
5-fold cross-validation score: 81.17%

trying model with 120 estimators...
Model accuracy on single test set split : 86.84%
5-fold cross-validation score: 83.16%

trying model with 130 estimators...
Model accuracy on single test set split : 85.53%
5-fold cross-validation score: 83.14%

trying model with 140 estimators...
Model accuracy on single test set split : 84.21%
5-fold cross-validation score: 82.48%

trying model with 150 estimators...
Model accuracy on single test set split : 84.21%
5-fold cross-validation score: 80.17%

trying model with 160 estimators...
Model accuracy on single test set split : 82.89%
5-fold cross-validation score: 80.83%

trying model with 170 estimators...
Model accuracy on single test set split : 86.84%
5-fold cross-validation score: 81.83%

trying m



Which model had the best cross-validation score?

This is usually a better indicator of a quality model than a single split accuracy score.

Rather than set up and track the results of these experiments manually, we can get Scikit-Learn to do the exploration for us.

Scikit-Learn's `sklearn.model_selection.GridSearchCV` is a way to search over a set of different hyperparameter values and *automatically* track which perform the best.

Let's test it!


In [22]:
# Another way to do it with GridSearchCV...
np.random.seed(42)
from sklearn.model_selection import GridSearchCV

# Define the parameters to search over in dictionary form
# (these can be any of your target model's hyperparameters) 
param_grid = {'n_estimators': [i for i in range(100,200, 10)]}

# setup the grid search
grid = GridSearchCV(estimator=RandomForestClassifier(),param_grid=param_grid,cv=5,verbose=1)

# fit the grid search to the data
grid.fit(X, y)

#find the best parameters
print(f"The best parameter values are : {grid.best_params_}")
print(f"With a score of : {grid.best_score_*100:.2f}%")


Fitting 5 folds for each of 10 candidates, totalling 50 fits
The best parameter values are : {'n_estimators': 120}
With a score of : 82.82%



We can extract the best model/estimator with the` best_estimator_` attribute.

In [23]:
# Set the model to the best estimator
clf = grid.best_estimator_
clf

RandomForestClassifier(n_estimators=120)

And now we've got the best cross-validated model, we can fit and score it on our original single train/test split of the data.

In [24]:
# fit the best model
clf = clf.fit(X_train, y_train)

# Find the best model scores on our single test split
# (note: this may be lower than the cross-validation score since it's only on one split of the data)
print(f'best model score on single split of the data: {clf.score(X_test, y_test)*100:.2f}%')

best model score on single split of the data: 86.84%



## 6. Save a model for someone else to use

When you've done a few experiments and you're happy with how your model is doing, you'll likely want someone else to be able to use it.

This may come in the form of a teammate or colleague trying to replicate and validate your results or through a customer using your model as part of a service or application you offer.

Saving a model also allows you to reuse it later without having to go through retraining it. Which is helpful, especially when your training times start to increase.

You can save a Scikit-Learn model using Python's in-built `pickle` module.


In [25]:
import pickle

# save an existing mode to file
pickle.dump(model, open("random_forest_model_1.pkl", "wb"))# "wb" stands for write binary


In [26]:
# Load a saved pickle model and evaluate it
loaded_pickle_model = pickle.load(open("random_forest_model_1.pkl", "rb"))
print(f"Loaded pickle model prediction score: {loaded_pickle_model.score(X_test, y_test) * 100:.2f}%")

Loaded pickle model prediction score: 86.84%




For larger models, it may be more efficient to use [Joblib].


In [27]:
from joblib import dump,load

# save a model using joblib
dump(model, "random_forest_model_1.joblib")

['random_forest_model_1.joblib']

In [28]:
# load a saved joblib model and evaluate it 
loaded_joblib_model = load("random_forest_model_1.joblib")
print(f"Loaded joblib model prediction score: {loaded_joblib_model.score(X_test, y_test) *100:.2f}%")

Loaded joblib model prediction score: 86.84%



## 1. Getting the data ready

Data doesn't always come ready to use with a Scikit-Learn machine learning model.

Three of the main steps you'll often have to take are:

  * Splitting the data into features (usually X) and labels (usually y).
  * Splitting the data into training and testing sets (and possibly a validation set).
  * Filling (also called imputing) or disregarding missing values.
  * Converting non-numerical values to numerical values (also call feature encoding).

Let's see an example.


In [29]:
# Splitting the data into X & y
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [30]:
# splitiing the data into features (X) and labels (y)
X = heart_disease.drop('target', axis=1)
X

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1        0   0     2  
2        2   0     2  
3        2   0     2  
4        2   0     2  
..     ...  ..   ...  
298      1   0     3  
299      1   0     3  
300      1   2     3  
301      1   1     3  
302      1   1     2  

[303 rows x 13 columns]

In [31]:
y = heart_disease['target']
y

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64

Now let's split our data into training and test sets, we'll use an 80/20 split (80% of samples for training and 20% of samples for testing).

In [32]:
# splitting the data into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2) 

# check the shapes of different data splits
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((242, 13), (61, 13), (242,), (61,))

In [33]:
# 80% of data is being used for the training set (the model will learn patterns on these samples)
X.shape[0]* 0.8

242.4

In [34]:

# And 20% of the data is being used for the testing set (the model will be evaluated on these samples)
X.shape[0] * 0.2

60.6


## 1.1 Make sure it's all numerical

Computers love numbers.

So one thing you'll often have to make sure of is that your datasets are in numerical form.

This even goes for datasets which contain non-numerical features that you may want to include in a model.

For example, if we were working with a car sales dataset, how might we turn features such as `Make` and `Colour` into numbers?

Let's figure it out.

First, we'll import the car-sales-extended.csv dataset.


In [35]:
# import extended car-sales file
car_sales = pd.read_csv("../data/car-sales-extended.csv")
car_sales

Make Colour  Odometer (KM)  Doors  Price
0     Honda  White          35431      4  15323
1       BMW   Blue         192714      5  19943
2     Honda  White          84714      4  28343
3    Toyota  White         154365      4  13434
4    Nissan   Blue         181577      3  14043
..      ...    ...            ...    ...    ...
995  Toyota  Black          35820      4  32042
996  Nissan  White         155144      3   5716
997  Nissan   Blue          66604      4  31570
998   Honda  White         215883      4   4001
999  Toyota   Blue         248360      4  12732

[1000 rows x 5 columns]

In [36]:
car_sales.dtypes

Make             object
Colour           object
Odometer (KM)     int64
Doors             int64
Price             int64
dtype: object



Notice the Make and Colour features are of `dtype=object` (they're strings) where as the rest of the columns are of dtype=int64.

If we want to use the Make and Colour features in our model, we'll need to figure out how to turn them into numerical form.


In [37]:
# Split into X & y and train/test
from sklearn.model_selection import train_test_split

X = car_sales.drop("Price", axis=1)
y = car_sales["Price"]

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2)

In [38]:
X_train.head()

Make Colour  Odometer (KM)  Doors
417   Honda   Blue          37625      4
666   Honda  White         211154      4
653     BMW   Blue          40774      5
822  Nissan  White         114699      4
762  Toyota  White         147230      4

In [39]:
# Try to predict with random forest on price column (doesn't work)
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor()
model.fit(X_train, y_train)
model.score(X_test, y_test)

ValueError: could not convert string to float: 'Honda'

Oh no! We get a another ValueError (some of data is in string format rather than numerical format).

Machine learning models prefer to work with numbers than text.

So we'll have to convert the non-numerical features into numbers first.

The process of turning categorical features into numbers is often referred to as encoding.

Scikit-Learn has a fantastic in-depth guide on Encoding categorical features.

But let's look at one of the most straightforward ways to turn categorical features into numbers, one-hot encoding.

In machine learning, one-hot encoding gives a value of 1 to the target value and a value of 0 to the other values.


In [40]:
# Import OneHotEncoder and ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# define the catergorical features to transform
categorical_features = ["Make", "Colour", "Doors"]

# create an instance of OneHotEncoder
one_hot = OneHotEncoder()

# Create an instance of ColumnTransformer
transformer = ColumnTransformer([("one_hot", #name
                                  one_hot, # transformer
                                  categorical_features)], # columns to transform
                                  remainder="passthrough") # what to do whith the rest

# 5. Turn the categorical features into numbers (this will return an array-like sparse matrix, not a DataFrame)
transformed_X = transformer.fit_transform(X)
transformed_X

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 3.54310e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        1.00000e+00, 1.92714e+05],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 8.47140e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 6.66040e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.15883e+05],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.48360e+05]])



**Note:** You might be thinking why we considered `Doors` as a categorical variable. Which is a good question considering Doors is already numerical. Well, the answer is that Doors could be either numerical or categorical. However, I've decided to go with categorical, since where I'm from, number of doors is often a different category of car. For example, you can shop for 4-door cars or shop for 5-door cars (which always confused me since where's the 5th door?). However, you could experiment with treating this value as numerical or categorical, training a model on each, and then see how each model performs.

Woah! Looks like our samples are all numerical, what did our data look like previously?


In [41]:
X.head()

Make Colour  Odometer (KM)  Doors
0   Honda  White          35431      4
1     BMW   Blue         192714      5
2   Honda  White          84714      4
3  Toyota  White         154365      4
4  Nissan   Blue         181577      3

In [42]:
# View first transformed sample
transformed_X[0]

array([0.0000e+00, 1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
       0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00,
       1.0000e+00, 0.0000e+00, 3.5431e+04])

In [43]:
# View original first sample
X.iloc[0]

Make             Honda
Colour           White
Odometer (KM)    35431
Doors                4
Name: 0, dtype: object


### 1.1.1 Nuemrically encoding data with pandas

Another way we can numerically encode data is directly with pandas.

We can use the pandas.get_dummies() (or pd.get_dummies() for short) method and then pass it our target columns.

In return, we'll get a one-hot encoded version of our target columns.

Let's remind ourselves of what our DataFrame looks like.


In [44]:
# View head of original DataFrame
car_sales.head()

Make Colour  Odometer (KM)  Doors  Price
0   Honda  White          35431      4  15323
1     BMW   Blue         192714      5  19943
2   Honda  White          84714      4  28343
3  Toyota  White         154365      4  13434
4  Nissan   Blue         181577      3  14043

In [45]:
# One-hot encode categorical variables
categorical_variables = ["Make", "Colour", "Doors"]
dummies = pd.get_dummies(data=car_sales[categorical_variables])
dummies

Doors  Make_BMW  Make_Honda  Make_Nissan  Make_Toyota  Colour_Black  \
0        4     False        True        False        False         False   
1        5      True       False        False        False         False   
2        4     False        True        False        False         False   
3        4     False       False        False         True         False   
4        3     False       False         True        False         False   
..     ...       ...         ...          ...          ...           ...   
995      4     False       False        False         True          True   
996      3     False       False         True        False         False   
997      4     False       False         True        False         False   
998      4     False        True        False        False         False   
999      4     False       False        False         True         False   

     Colour_Blue  Colour_Green  Colour_Red  Colour_White  
0          False         False       False          True  
1           True         False       False         False  
2          False         False       False          True  
3          False         False       False          True  
4           True         False       False         False  
..           ...           ...         ...           ...  
995        False         False       False         False  
996        False         False       False          True  
997         True         False       False         False  
998        False         False       False          True  
999         True         False       False         False  

[1000 rows x 10 columns]



Nice!

Notice how there's a new column for each categorical option (e.g. Make_BMW, Make_Honda, etc).

But also notice how it also missed the `Doors` column?

This is because Doors is already numeric, so for pd.get_dummies() to work on it, we can `change it to type object`.

By default, pd.get_dummies() also turns all of the values to bools (True or False).

We can get the returned values as `0 or 1` by setting `dtype=float`.


In [46]:
# Have to convert doors to object for dummies to work on it...
car_sales["Doors"] = car_sales["Doors"].astype(object)
dummies = pd.get_dummies(data=car_sales[["Make", "Colour", "Doors"]],
                         dtype=float)
dummies

Make_BMW  Make_Honda  Make_Nissan  Make_Toyota  Colour_Black  \
0         0.0         1.0          0.0          0.0           0.0   
1         1.0         0.0          0.0          0.0           0.0   
2         0.0         1.0          0.0          0.0           0.0   
3         0.0         0.0          0.0          1.0           0.0   
4         0.0         0.0          1.0          0.0           0.0   
..        ...         ...          ...          ...           ...   
995       0.0         0.0          0.0          1.0           1.0   
996       0.0         0.0          1.0          0.0           0.0   
997       0.0         0.0          1.0          0.0           0.0   
998       0.0         1.0          0.0          0.0           0.0   
999       0.0         0.0          0.0          1.0           0.0   

     Colour_Blue  Colour_Green  Colour_Red  Colour_White  Doors_3  Doors_4  \
0            0.0           0.0         0.0           1.0      0.0      1.0   
1            1.0           0.0         0.0           0.0      0.0      0.0   
2            0.0           0.0         0.0           1.0      0.0      1.0   
3            0.0           0.0         0.0           1.0      0.0      1.0   
4            1.0           0.0         0.0           0.0      1.0      0.0   
..           ...           ...         ...           ...      ...      ...   
995          0.0           0.0         0.0           0.0      0.0      1.0   
996          0.0           0.0         0.0           1.0      1.0      0.0   
997          1.0           0.0         0.0           0.0      0.0      1.0   
998          0.0           0.0         0.0           1.0      0.0      1.0   
999          1.0           0.0         0.0           0.0      0.0      1.0   

     Doors_5  
0        0.0  
1        1.0  
2        0.0  
3        0.0  
4        0.0  
..       ...  
995      0.0  
996      0.0  
997      0.0  
998      0.0  
999      0.0  

[1000 rows x 12 columns]

Now you might be wondering...

Should you use Scikit-Learn or pandas for turning data into numerical form?

And the answer is either.

But as a rule of thumb:

    If you're performing quick data analysis and running small modelling experiments, use pandas as it's generally quite fast to get up and running.
    If you're performing a larger scale modelling experiment or would like to put your data processing steps into a production pipeline, I'd recommend leaning towards Scikit-Learn, specifically a Scikit-Learn Pipeline (chaining together multiple estimator/modelling steps).

Since we've turned our data into numerical form, how about we try and fit our model again?

Let's recreate a train/test split except this time we'll use transformed_X instead of X.

In [47]:
np.random.seed(42)

# create train and test splits with transformed_X
X_train, X_test, y_train, y_test = train_test_split(transformed_X, y, test_size=0.2)

# Create the model instance
model = RandomForestRegressor()

# Fit the model on the numerical data (this errored before since our data wasn't fully numeric)
model.fit(X_train, y_train)

# Score the model (returns r^2 metric by default, also called coefficient of determination, higher is better)
model.score(X_test, y_test)

0.3235867221569877


## 1.2 What if there were missing values in the data?

Holes in the data means holes in the patterns your machine learning model can learn.

Many machine learning models don't work well or produce errors when they're used on datasets with missing values.

A missing value can appear as a blank, as a NaN or something similar.

There are two main options when dealing with missing values:

    Fill them with some given or calculated value (imputation) - For example, you might fill missing values of a numerical column with the mean of all the other values. The practice of calculating or figuring out how to fill missing values in a dataset is called imputing. For a great resource on imputing missing values, I'd recommend refering to the Scikit-Learn user guide.
    Remove them - If a row or sample has missing values, you may opt to remove them from your dataset completely. However, this potentially results in using less data to build your model.

    Note: Dealing with missing values differs from problem to problem, meaning there's no 100% best way to fill missing values across datasets and problem types. It will often take careful experimentation and practice to figure out the best way to deal with missing values in your own datasets.

To practice dealing with missing values, let's import a version of the car_sales dataset with several missing values (namely car-sales-extended-missing-data.csv).


In [48]:
# Import car sales dataframe with missing values
car_sales_missing = pd.read_csv("../data/car-sales-extended-missing-data.csv") # load from local directory
car_sales_missing.head(10)


Make Colour  Odometer (KM)  Doors    Price
0   Honda  White        35431.0    4.0  15323.0
1     BMW   Blue       192714.0    5.0  19943.0
2   Honda  White        84714.0    4.0  28343.0
3  Toyota  White       154365.0    4.0  13434.0
4  Nissan   Blue       181577.0    3.0  14043.0
5   Honda    Red        42652.0    4.0  23883.0
6  Toyota   Blue       163453.0    4.0   8473.0
7   Honda  White            NaN    4.0  20306.0
8     NaN  White       130538.0    4.0   9374.0
9   Honda   Blue        51029.0    4.0  26683.0

In [49]:
# Get the sum of all missing values
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64



Hmm... seems there's about 50 or so missing values per column.

How about we try and split the data into features and labels, then convert the categorical data to numbers, then split the data into training and test and then try and fit a model on it (just like we did before)?


In [50]:
# Create features
X_missing = car_sales_missing.drop("Price", axis=1)
print(f"Number of missing X values :\n{X_missing.isna().sum()}")

Number of missing X values :
Make             49
Colour           50
Odometer (KM)    50
Doors            50
dtype: int64


In [51]:
# Create labels
y_missing = car_sales_missing["Price"]
print(f"Number of missing y values: {y_missing.isna().sum()}")

Number of missing y values: 50




Now we can convert the categorical columns into one-hot encodings (just as before).


In [52]:
# Let's convert the categorical columns to one hot encoded (code copied from above)
# Turn the categories (Make and Colour) into numbers
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make", "Colour", "Doors"]

one_hot = OneHotEncoder()

transformer = ColumnTransformer([("one_hot", 
                                  one_hot, 
                                  categorical_features)],
                                remainder="passthrough",
                                sparse_threshold=0) # return a sparse matrix or not

transformed_X_missing = transformer.fit_transform(X_missing)
transformed_X_missing

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        0.00000e+00, 3.54310e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 1.92714e+05],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        0.00000e+00, 8.47140e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 0.00000e+00,
        0.00000e+00, 6.66040e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        0.00000e+00, 2.15883e+05],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        0.00000e+00, 2.48360e+05]])

Finally, let's split the missing data samples into train and test sets and then try to fit and score a model on them.

In [53]:
# Split data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(transformed_X_missing,
                                                    y_missing,
                                                    test_size=0.2)

# fit and score a model
model = RandomForestRegressor()
model.fit(X_train, y_test)
model.score(X_test, y_test)

ValueError: Input y contains NaN.

Ahh... dam! Another ValueError (our input data contains missing values).

`ValueError: Input y contains NaN.`

Looks like the model we're trying to use doesn't work with missing values.

When we try to fit it on a dataset with missing samples, Scikit-Learn produces an error similar to:

ValueError: Input X contains NaN. RandomForestRegressor does not accept missing values encoded as NaN natively...

Looks like if we want to use RandomForestRegressor, we'll have to either fill or remove the missing values.
Note: Scikit-Learn does have a list of models which can handle NaNs or missing values directly.

Such as, `sklearn.ensemble.HistGradientBoostingClassifier` or sklearn.ensemble.HistGradientBoostingRegressor.

As an experiment, you may want to try the following:


from sklearn.ensemble import HistGradientBoostingRegressor

```
# Try a model that can handle NaNs natively
nan_model = HistGradientBoostingRegressor()
nan_model.fit(X_train, y_train)
nan_model.score(X_test, y_test)
```

Let's
see what values are missing again.


In [54]:
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64


## 1.2.1 Fill missing data with pandas

Let's see how we might fill missing values with pandas.

For categorical values, one of the simplest ways is to fill the missing fields with the string "missing".

We could do this for the Make and Colour features.

As for the Doors feature, we could use "missing" or we could fill it with the most common option of 4.

With the Odometer (KM) feature, we can use the mean value of all the other values in the column.

And finally, for those samples which are missing a Price value, we can remove them (since Price is the target value, removing probably causes less harm than imputing, however, you could design an experiment to test this).

In summary:
```
Column/Feature 	Fill missing value with
Make 	"missing"
Colour 	"missing"
Doors 	4 (most common value)
Odometer (KM) 	mean of Odometer (KM)
Price (target) 	NA, remove samples missing Price
```
We can use the pandas method `fillna(value="missing", inplace=True)` to fill all the missing values with the string "missing".


In [55]:
# Fill the missing values in the Make column
# Note: In previous versions of pandas, inplace=True was possible, however this will be changed in a future version, can use reassignment instead.
# car_sales_missing["Make"].fillna(value="missing", inplace=True)

car_sales_missing["Make"] = car_sales_missing["Make"].fillna(value="missing")

In [56]:
# Fill the missing values in the Make column
# Note: In previous versions of pandas, inplace=True was possible, however this will be changed in a future version, can use reassignment instead.

# fill th colour column
car_sales_missing["Colour"] = car_sales_missing["Colour"].fillna(value="missing")

In [57]:
car_sales_missing.isna().sum()

Make              0
Colour            0
Odometer (KM)    50
Doors            50
Price            50
dtype: int64



Wonderful! We're making some progress.

Now let's fill the Doors column with 4 (the most common value), this is the same as filling it with the median or mode of the Doors column.


In [58]:
# Find the most common value of the Doors column
car_sales_missing["Doors"].value_counts()

Doors
4.0    811
5.0     75
3.0     64
Name: count, dtype: int64

In [59]:
# Fill the Doors column with the most common value
car_sales_missing["Doors"] = car_sales_missing["Doors"].fillna(value=4)

In [60]:
car_sales_missing.isna().sum()

Make              0
Colour            0
Odometer (KM)    50
Doors             0
Price            50
dtype: int64

In [61]:
# Fill the Odometer (Km) column
# Old:# Old: car_sales_missing["Odometer (KM)"].fillna(value=car_sales_missing["Odometer (KM)"].mean(), inplace=True)
car_sales_missing["Odometer (KM)"] = car_sales_missing["Odometer (KM)"].fillna(value=car_sales_missing["Odometer (KM)"].mean())



How many missing values do we have now?

In [62]:
# Check the number of missing values
car_sales_missing.isna().sum()

Make              0
Colour            0
Odometer (KM)     0
Doors             0
Price            50
dtype: int64



Woohoo! That's looking a lot better.

Finally, we can remove the rows which are missing the target value Price.

    Note: Another option would be to impute the Price value with the mean or median or some other calculated value (such as by using similar cars to estimate the price), however, to keep things simple and prevent introducing too many fake labels to the data, we'll remove the samples missing a Price value.

We can remove rows with missing values in place from a pandas DataFrame with the `pandas.DataFrame.dropna(inplace=True)` method.


In [63]:
# Remove rows with missing Price labels
car_sales_missing.dropna(inplace=True)

In [64]:
# Check the number of missing values
car_sales_missing.isna().sum()

Make             0
Colour           0
Odometer (KM)    0
Doors            0
Price            0
dtype: int64

Since we removed samples missing a Price value, there's now less overall samples but none of them have missing values.

In [65]:
# Check the number of total samples (previously was 1000)
len(car_sales_missing)

950



Can we fit a model now?

Let's try!

First we'll create the features and labels.

Then we'll convert categorical variables into numbers via one-hot encoding.

Then we'll split the data into training and test sets just like before.

Finally, we'll try to fit a `RandomForestRegressor()` model to the newly filled data.

In [66]:
# Create features
X_missing = car_sales_missing.drop("Price", axis =1)
print(f"Number of missing X values:\n{X_missing.isna().sum()}")

# create labels
y_missing = car_sales_missing["Price"]
print(f"Number of missing y values : {y_missing.isna().sum()}")

Number of missing X values:
Make             0
Colour           0
Odometer (KM)    0
Doors            0
dtype: int64
Number of missing y values : 0


In [67]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make", "Colour", "Doors"]

one_hot = OneHotEncoder()
transformer = ColumnTransformer([("one_hot",
                                  one_hot,
                                  categorical_features)],
                                remainder="passthrough",
                                sparse_threshold=0) # return a sparse matrix or not

transformed_X_missing = transformer.fit_transform(X_missing)
transformed_X_missing

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 3.54310e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        1.00000e+00, 1.92714e+05],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 8.47140e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 6.66040e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.15883e+05],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.48360e+05]])

In [68]:
# Split data into training and test sets
np.random.seed(42)
X_train, X_test, y_train, y_test = train_test_split(transformed_X_missing,y_missing,test_size=0.2) 

# Fit and score a model
model = RandomForestRegressor()
model.fit(X_train, y_train)
model.score(X_test, y_test)

0.22011714008302485


Fantastic!!!

Looks like filling the missing values with pandas worked!

Our model can be fit to the data without issues.


## 1.2.2 Filling missing data and transforming categorical data with Scikit-Learn

Now we've filled the missing columns using pandas functions, you might be thinking, "Why pandas? I thought this was a Scikit-Learn introduction?".

Not to worry, Scikit-Learn provides a class called `sklearn.impute.SimpleImputer()` which allows us to do a similar thing.

SimpleImputer() transforms data by filling missing values with a given strategy parameter.

And we can use it to fill the missing values in our DataFrame as above.

At the moment, our dataframe has no mising values.


In [69]:
car_sales_missing.isna().sum()

Make             0
Colour           0
Odometer (KM)    0
Doors            0
Price            0
dtype: int64

Let's reimport it so it has missing values and we can fill them with Scikit-Learn.

In [70]:
# Reimport the DataFrame (so that all the missing values are back)
car_sales_missing = pd.read_csv("../data/car-sales-extended-missing-data.csv")
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64

To begin, we'll remove the rows which are missing a Price value

In [71]:
# Drop the rows with missing in the price column
car_sales_missing.dropna(subset=["Price"], inplace=True)



Now there are no rows missing a Price value.


In [72]:
car_sales_missing.isna().sum()

Make             47
Colour           46
Odometer (KM)    48
Doors            47
Price             0
dtype: int64

Since we don't have to fill any Price values, let's split our data into features (X) and labels (y).
We'll also split the data into training and test sets.

In [73]:
# split into X and y
X = car_sales_missing.drop("Price", axis=1)
y = car_sales_missing["Price"]

# Split data into train and test
np.random.seed(42)
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2)


**Note:** We've split the data into train & test sets here first to perform filling missing values on them separately. This is best practice as the test set is supposed to emulate data the model has never seen before. For categorical variables, it's generally okay to fill values across the whole dataset. However, for numerical vairables, you should only fill values on the test set that have been computed from the training set.

Training and test sets created!

Let's now setup a few instances of SimpleImputer() to fill various missing values.

We'll use the following strategies and fill values:

* For categorical columns (Make, Colour), strategy="constant", fill_value="missing" (fill the missing samples with a consistent value of "missing".
  * For the Door column, strategy="constant", fill_value=4 (fill the missing samples with a consistent value of 4).
  * For the numerical column (Odometer (KM)), strategy="mean" (fill the missing samples with the mean of the target column).
       * Note: There are more strategy and fill options in the SimpleImputer()

In [74]:
from sklearn.impute import SimpleImputer

# Create categorical variable imputer
cat_imputer = SimpleImputer(strategy="constant", fill_value="missing")

# create Door column imputer
door_imputer = SimpleImputer(strategy= "constant" , fill_value=4)

# Create Odometer (KM) column imputer
num_imputer = SimpleImputer(strategy="mean")

Imputer created

Now let's define which columns we'd like to imputer on

Because we'll need these shortly(i'll explain in the next text cell).

In [75]:
# Define different column features
categorical_features = ["Make", "Colour"]
door_feature = ["Doors"]
numerical_feature = ["Odometer (KM)"]

Columns defined!
Now how might we transform our columns?

Hint: we can use the sklearn.compose.ColumnTransformer class from Scikit-Learn, in a similar way to what we did before to get our data to all numeric values.

That's the reason we defined the columns we'd like to transform.

So we can use the `ColumnTransformer()` class.

ColumnTransformer() takes as input a list of tuples in the form (name_of_transform, transformer_to_use, columns_to_transform) specifying which columns to transform and how to transform them.

In [76]:
from sklearn.compose import ColumnTransformer

# create series of column trasforms to perform
imputer = ColumnTransformer([
    ("cat_imputer", cat_imputer, categorical_features),
    ("door_imputer", door_imputer, door_feature),
    ("num_imputer", num_imputer, numerical_feature)])



Nice!

The next step is to fit our ColumnTransformer() instance (imputer) to the training data and transform the testing data.

In other words we want to:

    Learn the imputation values from the training set.
    Fill the missing values in the training set with the values learned in 1.
    Fill the missing values in the testing set with the values learned in 1.

Why this way?m

In our case, we're not calculating many variables (except the mean of the Odometer (KM) column), however, remember that the test set should always remain as unseen data.

So **when filling values in the test set, they should only be with values calculated or imputed from the training sets.**

We can achieve steps 1 & 2 simultaneously with the ColumnTransformer.fit_transform() method (fit = find the values to fill, transform = fill them).

And then we can perform step 3 with the ColumnTransformer.transform() method (we only want to transform the test set, not learn different values to fill).


In [77]:
# find vaues to fill and transform training data
filled_X_train = imputer.fit_transform(X_train)

# fill values into the test set with values learned from the training set
filled_X_test = imputer.transform(X_test)

# check filled X_train
filled_X_train

array([['Honda', 'White', 4.0, 71934.0],
       ['Toyota', 'Red', 4.0, 162665.0],
       ['Honda', 'White', 4.0, 42844.0],
       ...,
       ['Toyota', 'White', 4.0, 196225.0],
       ['Honda', 'Blue', 4.0, 133117.0],
       ['Honda', 'missing', 4.0, 150582.0]], dtype=object)

Wonderful!

Let's now turn out filled_X_train and `filled_X_test` arrays into Dataframes to inspect thier missing values

In [78]:
# Get our transformed data arry''s back into DataFrame's
filled_X_train_df = pd.DataFrame(filled_X_train, columns=["Make", "Colour", "Doors", "Odometer (KM)"])

filled_X_test_df = pd.DataFrame(filled_X_test, columns=["Make", "Colour", "Doors", "Odometer (KM)"])

#check missing data in training set
filled_X_train_df.isna().sum()

Make             0
Colour           0
Doors            0
Odometer (KM)    0
dtype: int64

In [79]:
# check to see the original.. still misssing values
car_sales_missing.isna().sum()

Make             47
Colour           46
Odometer (KM)    48
Doors            47
Price             0
dtype: int64



Perfect!

No more missing values!

But wait...

Is our data all numerical?


In [80]:
filled_X_train_df.head()

Make Colour Doors Odometer (KM)
0   Honda  White   4.0       71934.0
1  Toyota    Red   4.0      162665.0
2   Honda  White   4.0       42844.0
3   Honda  White   4.0      195829.0
4   Honda   Blue   4.0      219217.0



Ahh... looks like our Make and Colour columns are still strings.

Let's one-hot encode them along with the Doors column to make sure they're numerical, just as we did previously.


In [81]:
# Now let's one hot encode the features with the same code as before
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

categorical_features = ["Make", "Colour", "Doors"]

one_hot = OneHotEncoder()

transformer = ColumnTransformer([("one_hot",
                                  one_hot,
                                  categorical_features)],
                                  remainder="passthrough",
                                sparse_threshold=0) # return a sparse matrixe or not

# fill train and test values separately
transformed_X_train = transformer.fit_transform(filled_X_train_df)
transformed_X_test = transformer.transform(filled_X_test_df)

# check transformed and filled X_train
transformed_X_train


array([[0.0, 1.0, 0.0, ..., 1.0, 0.0, 71934.0],
       [0.0, 0.0, 0.0, ..., 1.0, 0.0, 162665.0],
       [0.0, 1.0, 0.0, ..., 1.0, 0.0, 42844.0],
       ...,
       [0.0, 0.0, 0.0, ..., 1.0, 0.0, 196225.0],
       [0.0, 1.0, 0.0, ..., 1.0, 0.0, 133117.0],
       [0.0, 1.0, 0.0, ..., 1.0, 0.0, 150582.0]], dtype=object)



Nice!

Now our data is:

* All numerical
* No missing values

Let's try and fit a model!


In [82]:
# Now we've transformed X, let's see if we can fit a model
np.random.seed(42)
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor()

# Make sure  to use the transformed data (filled and one-hot encoded X data)
model.fit(transformed_X_train, y_train)
model.score(transformed_X_test, y_test)

0.21229043336119102



You might have noticed this result is slightly different to before.

Why do you think this is?

It's because we've created our training and testing sets differently.

We split the data into training and test sets before filling the missing values.

Previously, we did the reverse, filled missing values before splitting the data into training and test sets.

Doing this can lead to information from the training set leaking into the testing set.

Remember, one of the most important concepts in machine learning is making sure your model doesn't see any testing data before evaluation.

We'll keep practicing but for now, some of the main takeaways are:

* Keep your training and test sets separate.
* Most datasets you come across won't be in a form ready to immediately start using them with machine learning models. And some may take more preparation than others to get ready to use.
* For most machine learning models, your data has to be numerical. This will involve converting whatever you're working with into numbers. This process is often referred to as **feature engineering** or **feature encoding**.
* Some machine learning models aren't compatible with missing data. The process of filling missing data is referred to as data imputation.



## 2. Choosing the right estimator/algorithm for your problem

Once you've got your data ready, the next step is to choose an appropriate machine learning algorithm or model to find patterns in your data.

Some things to note:

* Scikit-Learn refers to machine learning models and algorithms as estimators.
* **Classification problem** - predicting a category (heart disease or not).
  * Sometimes you'll see clf (short for classifier) used as a classification estimator instance's variable name.
* **Regression problem** - predicting a number (selling price of a car).
* **Unsupervised problem** (data with no labels) - clustering (grouping unlabelled samples with other similar unlabelled samples).

If you know what kind of problem you're working with, one of the next places you should look at is the -https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html.

This cheatsheet gives you a bit of an insight into the algorithm you might want to use for the problem you're working on.

It's important to remember, you don't have to explicitly know what each algorithm is doing on the inside to start using them.

If you start to apply different algorithms but they don't seem to be working (not performing as well as you'd like), that's when you'd start to look deeper into each one.

Let's check out the cheatsheet and follow it for some of the problems we're working on.

You can see it's split into four main categories. Regression, classification, clustering and dimensionality reduction. Each has their own different purpose but the Scikit-Learn team has designed the library so the workflows for each are relatively similar.
## 2.1 Picking a machine learning model for a regression problem

Let's start with a regression problem (trying to predict a number). We'll use the California Housing dataset built into Scikit-Learn's datasets module.

The goal of the California Housing dataset is to predict a given district's median house value (in hundreds of thousands of dollars) on things like the age of the home, the number of rooms, the number of bedrooms, number of people living the home and more.


In [83]:
# Get California Housing dataset
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
housing; # gets downloade as dictionary

since it's dictionary let's turn it into a DataFrame so we can inspect it better

In [85]:
print(housing)

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
          37.88      , -122.23      ],
       [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
          37.86      , -122.22      ],
       [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
          37.85      , -122.24      ],
       ...,
       [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
          39.43      , -121.22      ],
       [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
          39.43      , -121.32      ],
       [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
          39.37      , -121.24      ]]), 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]), 'frame': None, 'target_names': ['MedHouseVal'], 'feature_names': ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude'], 'DESCR': '.. _california_housing_dataset:\n\nCalifornia Housing dataset\n-

In [86]:
housing_df = pd.DataFrame(housing["data"], columns=housing["feature_names"])
housing_df["target"] = pd.Series(housing["target"])
housing_df.head()
                          

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

In [87]:
# How many samples?
len(housing_df)

20640

Beautiful, our goal here is to use the feature columns, such as:

* MedInc - median income in block group
* HouseAge - median house age in block group
* AveRooms - average number of rooms per household
* AveBedrms - average number of bedrooms per household

To predict the target column which expresses the median house value for specfici California districts in hundreds of thousands of dollars (e.g. 4.526 = $452,600).

In essence, each row is a different district in California (the data) and we're trying to build a model to predict the median house value in that distract (the target/label) given a series of attributes about the houses in that district.

Since we have data and labels, this is a supervised learning problem.

And since we're trying to predict a number, it's a regression problem.

**Following the map through, knowing what we know, it suggests we try RidgeRegression**

In [90]:
# Import the Ridge model class from the linear_model module
from sklearn.linear_model import Ridge

# Setup random seed
np.random.seed(42)

# Split the data into features (x) and labels (y)
X = housing_df.drop("target", axis=1)
y = housing_df["target"]

# split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2)

# instintiate and fit the model (on the training set )
model = Ridge()
model.fit(X_train, y_train)

# check the score of the model (on the test set )
# the default score()  metric of regression algorithms in R²
model.score(X_test, y_test)

0.5758549611440125

What if RidgeRegression didn't work? Or what if we wanted to improve our results?

Following the diagram, the next step would be to try EnsembleRegressors.

Ensemble is another word for multiple models put together to make a decision.

One of the most common and useful ensemble methods is the Random Forest. Known for its fast training and prediction times and adaptibility to different problems.

The basic premise of the Random Forest is to combine a number of different decision trees, each one random from the other and make a prediction on a sample by averaging the result of each decision tree.

An in-depth discussion of the Random Forest algorithm is beyond the scope of this notebook but if you're interested in learning more, An Implementation and Explanation of the Random Forest in Python by Will Koehrsen is a great read.

Since we're working with regression, we'll use Scikit-Learn's `RandomForestRegressor.`

We can use the exact same workflow as above. Except for changing the model.

In [91]:
# import the RandomForestRegressor model class from the ensemble module
from sklearn.ensemble import RandomForestRegressor

# setup random seed
np.random.seed(42)

# split the data into features (x) and labels (y)
X = housing_df.drop("target", axis=1)
y = housing_df["target"]

# split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# instintiate and fit the model (on the training set)
model = RandomForestRegressor()
model.fit(X_train, y_train)

# Check the score of the model (on test set)
# the default score metric of regression algorthms is R²
model.score(X_test, y_test)

0.8066196804802649

We get a good boost in score on the test set by changing the model.

This is another incredibly important concept in machine learning, if at first something doesn't achieve what you'd like, experiment, experiment, experiment!

At first, the Scikit-Learn algorithm diagram can seem confusing.


## 2.2 Picking a machine learning model for a classification problem

Now, let's check out the choosing process for a classification problem.

Say you were trying to predict whether or not a patient had heart disease based on their medical records.

The dataset in ../data/heart-disease.csv (or at heart-disease.csv) contains data for just that problem.


In [ ]:
heart_disease = pd.read_csv("../data/heart-disease.csv") # load from local directory
